tree: add resolver argument to MakeSemaContext

yuzefovich opened the pull request:
This commit introduces a resolver argument to `MakeSemaContext` (which can be left nil) that will be tried against three "resolver" interfaces stored in the sema context. This should make it less likely in the future that we'd run into a bug similar to the one fixed in 108990dbb102cdc148085b50ef70be5b3c567b2c.

Release note: None

Epic: None

yuzefovich committed:
tree: add resolver argument to MakeSemaContext

This commit introduces a resolver argument to `MakeSemaContext` (which
can be left nil) that will be tried against three "resolver" interfaces
stored in the sema context. This should make it less likely in the
future that we'd run into a bug similar to the one fixed in 108990dbb102cdc148085b50ef70be5b3c567b2c.

Release note: None

yuzefovich committed:
tree: remove seemingly redundant fresh SemaCtx when type checking const

This was introduced long time ago in 2e0f850ef5a6671c4d42a330195e31cc0fb71fcb
and I don't see a reason why we can't use the sema context we have in scope.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/122315)
<!-- Reviewable:end -->

yuzefovich committed:
sql: pass sema ctx from the job exec ctx for stats

This commit plumbs propagation of the sema context from the JobExecCtx
to the stats job. Previously, we created a fresh one and that was
breaking the following commit (I didn't quite understand why, probably
due to the way we hide `planner` behind the `JobExecContext` interface).

It also adjusts the type-checking logic a bit to remove an assumption
that non-nil sema context has SearchPath set (also needed by the
following commit).

Release note: None

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice, thank you!

> Reviewed 1 of 1 files at r1, 4 of 4 files at r2, 84 of 84 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/122315#-:-Nvsayst2QOe0XDB8MmB:b-xkcb0o)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
